Change command alignment verification to pass or warn

lukpueh opened the pull request:
- ~~Adds 3 custom class based exception and modifies functions in `verifiylib` and in `common.Signable` to use them.~~
- ~~Replaces use of `log` with a `print` statement for a user warning~~

This PR used to be `Add verifylib exceptions` but the focus changed during review so it was renamed. This PR now:
- changes command alignment verification (see https://github.com/in-toto/in-toto/pull/47/commits/d91479397964958bcfd357c68b6d08e713e3d283 for reasoning) + unittests
- Adds and uses custom `SignatureVerificationError`
- Adds and uses custom `LayoutExpiredError`

lukpueh committed:
Adds CommandAlignmentFailed + replaces log

- Adds class based Exception for command alignment fail
- Replaces relaxed alignment warning log with print
  This is only temporary (see inline comments)

lukpueh committed:
Adds LayoutExpiredError exception

- Replaces generic Exception with new LayoutExpiredError
  Is raised when the layout is expired

lukpueh committed:
Adds SignatureVerificationError

currently raised if a signable's signatures are verified and
- it has no signatures
- the verification signature does not have the right key
- the actual signature verification fails

Note:
ssl_commons has a BadSignature exception which could maybe be
used instead

SantiagoTorres commented on toto/exceptions.py:
```diff
@@ -1,5 +1,15 @@
+
+class LayoutExpiredError(Error):
+  """Indicates that the layout expired. """
+  pass
```
Some have pass, some others don't. Is that intentional?

SantiagoTorres commented on toto/models/common.py:
```diff
@@ -29,6 +29,7 @@
 import attr
 import canonicaljson
 
+from ..exceptions import SignatureVerificationError
```
Does this need to be a relative import? or can we import toto.exceptions?

SantiagoTorres commented on toto/verifylib.py:
```diff
@@ -213,11 +213,15 @@ def verify_command_alignment(command, expected_command):
     if command[i] != expected_command[i]:
-      raise Exception("Command '%s' and expected command '%s' do not align" \
+      raise CommandAlignmentFailed(
+          "Command '%s' and expected command '%s' do not align" \
```
shoudln't we be moving to `format`?

SantiagoTorres replied:
Also, I'm not sure if you need the line splice since this is within parentheses.

SantiagoTorres commented on toto/verifylib.py:
```diff
@@ -213,11 +213,15 @@ def verify_command_alignment(command, expected_command):
+    # FIXME we don't want print statements in a library
+    # The warning will be obsolete with a fix for
+    # https://github.com/in-toto/in-toto/issues/46
+    print "Command '%s' and expected command '%s' do not fully align" \
```
can we add parentheses to it so we don't break when we  hit python 3?

lukpueh committed:
Replaces python's '%'-style formatting with format

- while we are working on this we might as well start
using "{}".format instead of "%s" % () as suggested in
https://github.com/in-toto/in-toto/issues/25
- also add parentheses to print statements as a step of achieving
python 3 compatibility
https://github.com/in-toto/in-toto/issues/23

lukpueh committed:
Removes relative imports where unneeded in common

We started using relative imports in interdependent
subpackages that can be used in multiple projects,
e.g. in-toto and tuf. Submodules that are specific
for a project can actually use absolute imports

lukpueh commented:
Thanks @SantiagoTorres. Addressed your suggestions in 8c33184 and 9247f33.

SantiagoTorres commented on toto/verifylib.py:
```diff
@@ -213,11 +213,15 @@ def verify_command_alignment(command, expected_command):
+    # FIXME we don't want print statements in a library
+    # The warning will be obsolete with a fix for
+    # https://github.com/in-toto/in-toto/issues/46
+    print("Command '{0}' and expected command '{1}' do not fully align"
```
I'm curious, why don't we use the logger instance as it was done before? 
`log.warning` sounded like a good idea.

lukpueh replied:
Hm. I talked to @vladimir-v-diaz about it. We agreed that the `logging` module should not be used to give direct feedback to the user, but rather to *log* errors/warnings/infos to some place (could be a file) intended for the developer/debugger.

vladimir-v-diaz replied:
We figured that if in-toto is going to be used as a library, the calling application probably doesn't want to see messages from the library printed/logged to the console (at least not by default).  For applications that would like to have these log statements show up on the console, perhaps an option can be added for them?

SantiagoTorres replied:
Ok, then it would make sense to add exceptions to pr #40 and let the user-facing tool handle them?

lukpueh replied:
So I re-thought all of this and summarized my conclusions in https://github.com/in-toto/in-toto/issues/6#issuecomment-263988714. Nevertheless, this still does not solve this particular issue. If we raise an exception instead of a warning here, the exception could be caught and translated to a warning in the user tool, but this would also abort the the layout verification, which we don't want.

lukpueh replied:
Solutions:
 - return a list of "pending command verification exceptions"  that can be raised, caught and logged once the entire verification has finished
 - just log the warning in the library

lukpueh committed:
Replaces command alignment exception with warning

In a previous commit we added a custom CommandAlignmentFailed
exception and replaced log with print in case of soft fail.
This commit removes the custom exception (1) and changes print back
to log (2) for the following reasons.

(1) We decided use log throughout the library also for user feedback
because it is more powerful and can easily be toggled
programatically, see discussion
https://github.com/in-toto/in-toto/issues/6#issuecomment-263988714

(2) We decided to never hard fail command alignment because the check
is of informational character rather than a security guarantee. See
https://github.com/in-toto/in-toto/issues/46

The commit also adds unit tests for command alignment

lukpueh commented:
This PR is starting to diverge from its original intent. The most recent commit drastically changes command alignment verification and adds unittests. The changes were triggered by a discussion over the introduced `CommandAlignmentFailed` Exception, which eventually was removed again in the most recent commit.

SantiagoTorres commented:
Yeah, so maybe we should close it and start over once we decide what's the best path forward?

lukpueh commented:
I don't think that's necessary. We could:
 - Rename the PR
 - Update the PR's description
 - (rename the branch?)

lukpueh committed:
Adds mock package to ci requirements, adds newline

We use mock package in unittests to see if the right warning is
logged when a ran command does not align with an expected command

Adds a newline at the end of exceptions.py
